homeassistant/helpers/network.py

```diff
@@ -5,8 +5,128 @@
 import yarl
 
 from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
 from homeassistant.loader import bind_hass
-from homeassistant.util.network import is_local
+from homeassistant.util import get_local_ip
+from homeassistant.util.network import is_ip_address, is_local, is_loopback
+
+
+class NoURLAvailableError(HomeAssistantError):
+    """An URL to the Home Assistant instance is not available."""
+
+
+@bind_hass
+@callback
+def async_get_url(
+    hass: HomeAssistant,
+    *,
+    require_ssl: bool = False,
+    require_standard_port: bool = False,
+    allow_local: bool = True,
+) -> Optional[str]:
+    """Get a URL to this instance."""
+    # Return internal URL when available and matches requested citeria
+    if allow_local and hass.config.internal_url:
+        internal_url = yarl.URL(hass.config.internal_url.rstrip("/"))
+        if (not require_ssl or internal_url.scheme == "https") and (
+            not require_standard_port or internal_url.is_default_port()
+        ):
+            return (
+                str(internal_url)
+                if not internal_url.is_default_port()
+                else str(internal_url.with_port(None))
+            )
+
+    # Return IP if local is allowed and it matches citeria requested.
+    # This cannot be used if SSL is required, as IP based will always result
+    # in a invalid SSL certificate.
+    if (
+        allow_local
+        and not require_ssl
+        and hass.config.api is not None
+        and not hass.config.api.use_ssl
+    ):
+        ip_url = yarl.URL.build(
+            scheme="http", host=get_local_ip(), port=hass.config.api.port
+        )
+        if not is_loopback(ip_address(ip_url.host)) and (
+            not require_standard_port or ip_url.is_default_port()
+        ):
+            return (
+                str(ip_url)
+                if not ip_url.is_default_port()
+                else str(ip_url.with_port(None))
+            )
+
+    # Return user set external URL if it matches the requested citeria
+    if hass.config.external_url:
+        external_url = yarl.URL(hass.config.external_url.rstrip("/"))
```

Comment:
I think that the rstrip should be done when we save it.

Reply:
That is not a real option, as it can be supplied by YAML configuration as well.

However, the URL normalizer handles this as well, so removed it here.